# Arulin Vs. Novadra: Clinical Trial Analysis

## Gather

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
patients = pd.read_csv("patients.csv")
treatments = pd.concat([pd.read_csv("treatments.csv"), pd.read_csv("treatments_cut.csv")])
reactions = pd.read_csv("adverse_reactions.csv")


In [3]:
patients.sample(5)

patient_id assigned_sex  given_name        surname  \
220         221         male      Mijael         Guerra   
184         185       female        Ásta    Grímsdóttir   
135         136         male  Willem-Jan  van der Lubbe   
490         491         male     Jackson        Addison   
90           91         male        Ingo         Rokavc   

                     address          city     state  zip_code        country  \
220    1271 Pretty View Lane       Oakland        CA   94612.0  United States   
184         1619 Melm Street  Ormond Beach        FL   32174.0  United States   
135      1717 Vineyard Drive     Cleveland        OH   44115.0  United States   
490       1160 Taylor Street  New Rochelle  New York   10801.0  United States   
90   4271 Cherry Ridge Drive       Buffalo  New York   14214.0  United States   

                                             contact  birthdate  weight  \
220  MijaelGuerraMoreno@teleworm.us+1 (707) 896-9250  1/25/1944   183.5   
184      AstaGrimsdottir@dayrep.com+1 (386) 989-0019   6/9/1999   128.9   
135      Willem-JanvanderLubbe@gustr.com440-385-5011   7/9/1941   152.9   
490           914-636-9304JacksonAddison@armyspy.com  5/29/1953   192.7   
90         IngoRokavc@superrito.com+1 (585) 902-9127  8/27/1977   192.3   

     height   bmi  
220      65  30.5  
184      65  21.4  
135      69  22.6  
490      69  28.5  
90       67  30.1

In [4]:
treatments.sample(5)

given_name   surname    auralin    novodra  hba1c_start  hba1c_end  \
36         chỉ       lâm  45u - 48u          -         7.68       7.24   
61        ella      lund  38u - 47u          -         7.57       7.21   
210    vanessa  ferguson  39u - 50u          -         7.97       7.58   
26     firenze     fodor          -  30u - 35u         7.89       7.55   
278     vallie    prince  31u - 38u          -         7.64       7.28   

     hba1c_change  
36            NaN  
61           0.36  
210           NaN  
26           0.34  
278          0.36

In [5]:
reactions.sample(5)

given_name   surname adverse_reaction
7       albinca   komavec     hypoglycemia
27       idalia     moore     hypoglycemia
8           noe    aranda     hypoglycemia
17  christopher  woodward           nausea
20     anenechi     chidi     hypoglycemia

## Assess

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
patients[["weight", "height", "bmi"]].describe()

weight      height         bmi
count  503.000000  503.000000  503.000000
mean   173.434990   66.634195   27.483897
std     33.916741    4.411297    5.276438
min     48.800000   27.000000   17.100000
25%    149.300000   63.000000   23.300000
50%    175.300000   67.000000   27.200000
75%    199.500000   70.000000   31.750000
max    255.900000   79.000000   37.700000

In [8]:
mask = patients["weight"] < 100
patients[mask]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

In [9]:
mask = patients["height"] < 50
patients[mask]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state  zip_code        country                          contact  birthdate  \
4    AL   36303.0  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  
4   192.3      27  26.1

In [10]:
patients.duplicated().sum()

0

In [11]:
patients.duplicated(subset=["surname"]).sum()

37

In [12]:
mask = patients.duplicated(subset=["surname"])
patients[mask]

patient_id assigned_sex  given_name      surname  \
26           27       female        Ella         Lund   
29           30         male        Jake     Jakobsen   
122         123         male         Bội           Tạ   
139         140       female     Novalie         Berg   
194         195         male        Urso       Aranda   
195         196       female      Lamara     Dratchev   
203         204       female         Mùi        Lương   
205         206         male       Breno      Correia   
207         208       female   Beatrycze      Woźniak   
229         230         male        John          Doe   
235         236       female     Justyna    Kowalczyk   
237         238         male        John          Doe   
244         245         male        John          Doe   
251         252         male        John          Doe   
277         278         male        John          Doe   
282         283       female       Sandy       Taylor   
295         296         male      Albino    Schiavone   
296         297       female         Chỉ          Lâm   
304         305       female       Signý  Grímsdóttir   
308         309         male        Daud    Batukayev   
317         318       female       Nancy       Parker   
322         323         male        Mười         Liễu   
336         337       female      Sharon      Collins   
350         351         male        Diệt          Bùi   
364         365       female  Chukwumoge    Ogochukwu   
407         408       female       Tegan      Johnson   
418         419         male      Mahmud      Kadyrov   
426         427         male     Rogelio       Taylor   
431         432       female     Manuela      Cindrić   
432         433       female       Karen     Jakobsen   
441         442         male        Luis        Silva   
449         450         male     Bjørnar       Nilsen   
461         462         male      Cannan      Cabrera   
469         470         male      Harley       Tucker   
482         483         male       Diogo        Souza   
496         497         male   Alexander       Hueber   
502         503         male         Pat      Gersten   

                      address              city       state  zip_code  \
26         1207 Garfield Road            Peoria          IL   61602.0   
29          648 Old Dear Lane       Port Jervis    New York   12771.0   
122        2389 Rubaiyat Road      Grand Rapids          MI   49503.0   
139          1275 Goldie Lane        Cincinnati          OH   45202.0   
194       1330 Lincoln Street  Hopewell  Mercer          NJ    8525.0   
195    3731 Swick Hill Street       New Orleans          LA   70113.0   
203        1778 Rodney Street         Harvester          MO   63301.0   
205          3539 Bottom Lane         Tonawanda    New York   14150.0   
207          182 Cross Street           Saginaw          MI   48607.0   
229           123 Main Street          New York          NY   12345.0   
235           3115 May Street      Pippa Passes          KY   41844.0   
237           123 Main Street          New York          NY   12345.0   
244           123 Main Street          New York          NY   12345.0   
251           123 Main Street          New York          NY   12345.0   
277           123 Main Street          New York          NY   12345.0   
282        2476 Fulton Street          Rainelle          WV   25962.0   
295    3113 Timber Ridge Road        Sacramento  California   95814.0   
296                       NaN               NaN         NaN       NaN   
304           3268 Karen Lane        Louisville          KY   40202.0   
308       153 Fieldcrest Road        Huntington    New York   11743.0   
317          4605 Hall Street         Las Vegas          NV   89110.0   
322         4346 Sussex Court           Chilton          TX   76632.0   
336   1240 Summit Park Avenue           Pontiac          MI   48342.0   
350    3788 Worthington Drive             Rylie          TX   75235.0   
364          

In [14]:
patients.duplicated(subset=["contact"]).sum()

19

In [19]:
mask = patients.duplicated(subset=["contact"], keep=False)
patients.dropna(subset=["contact"])[mask]

<ipython-input-19-2ffeb6f84185>:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  patients.dropna(subset=["contact"])[mask]


patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
97           98         male    Patrick   Gersten   2778 North Avenue   
131         132       female     Sandra    Taylor  2476 Fulton Street   
215         216         male       John       Doe     123 Main Street   
229         230         male       John       Doe     123 Main Street   
237         238         male       John       Doe     123 Main Street   
244         245         male       John       Doe     123 Main Street   
251         252         male       John       Doe     123 Main Street   
277         278         male       John       Doe     123 Main Street   
282         283       female      Sandy    Taylor  2476 Fulton Street   
502         503         male        Pat   Gersten   2778 North Avenue   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
97          Burr        NE   68324.0  United States   
131     Rainelle        WV   25962.0  United States   
215     New York        NY   12345.0  United States   
229     New York        NY   12345.0  United States   
237     New York        NY   12345.0  United States   
244     New York        NY   12345.0  United States   
251     New York        NY   12345.0  United States   
277     New York        NY   12345.0  United States   
282     Rainelle        WV   25962.0  United States   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
97         PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71   
131        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
215                 johndoe@email.com1234567890    1/1/1975   180.0      72   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
502        PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71   

      bmi  
24   24.4  
29   24.4  
97   19.3  
131  35.4  
215  24.4  
229  24.4  
237  24.4  
244  24.4  
251  24.4  
277  24.4  
282  35.4  
502  19.3

In [20]:
patients["state"].value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
Illinois      14
OH            14
Florida       13
MI            13
OK            13
LA            13
NJ            12
VA            11
MS            10
WI            10
IL            10
IN             9
AL             9
MN             9
TN             9
FL             9
NC             8
KY             8
WA             8
MO             7
KS             6
NV             6
ID             6
SC             5
CT             5
IA             5
Nebraska       4
AR             4
CO             4
AZ             4
RI             4
ND             4
ME             4
OR             3
WV             3
DE             3
MD             3
SD             3
VT             2
DC             2
NE             2
MT             2
NM             1
AK             1
NH             1
WY             1
Name: state, dtype: int64

In [23]:
patients.isnull().sum()

patient_id       0
assigned_sex     0
given_name       0
surname          0
address         12
city            12
state           12
zip_code        12
country         12
contact         12
birthdate        0
weight           0
height           0
bmi              0
dtype: int64

In [24]:
treatments.isnull().sum()

given_name        0
surname           0
auralin           0
novodra           0
hba1c_start       0
hba1c_end         0
hba1c_change    137
dtype: int64

In [25]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 21.9+ KB


In [26]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [27]:
treatments.duplicated().sum()

1

In [28]:
treatments.duplicated(subset=["surname"]).sum()

17

In [32]:
mask = treatments.duplicated(subset=["surname"], keep=False)
treatments[mask].sort_values("surname")

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
19            noe     aranda  26u - 34u          -         7.51       7.17   
132          urso     aranda  30u - 38u          -         7.60       7.16   
49        satsita  batukayev          -  42u - 42u         7.63       7.25   
226          daud  batukayev          -  37u - 30u         7.98       7.51   
199       novalie       berg          -  32u - 31u         7.85       7.49   
24           isac       berg  31u - 41u          -         9.68       9.29   
219          diệt        bùi          -  30u - 32u         9.11       8.76   
184          chân        bùi  31u - 42u          -         7.53       7.18   
13       bernarda    cindrić  40u - 49u          -         7.89       7.55   
150       manuela    cindrić  55u - 66u          -         8.07       7.76   
10         joseph        day  29u - 36u          -         7.70       7.19   
136        joseph        day  29u - 36u          -         7.70       7.19   
39         maryam   dratchev  37u - 45u          -         9.91       9.46   
167        lamara   dratchev  49u - 58u          -         7.75       7.40   
238          anja     hueber          -  31u - 32u         8.82       8.36   
212     alexander     hueber  35u - 38u          -         7.71       7.35   
66          jakob   jakobsen          -  28u - 26u         7.96       7.51   
90          karen   jakobsen  34u - 42u          -         7.59       7.25   
183         tegan    johnson          -  34u - 34u         7.79       7.43   
66         louise    johnson  32u - 42u          -         7.63       7.32   
236         hường        lâm  40u - 49u          -         7.66       7.29   
36            chỉ        lâm  45u - 48u          -         7.68       7.24   
3             thể      lương          -  61u - 64u         7.64       7.22   
108           mùi      lương          -  52u - 50u         7.74       7.32   
213       cecilie     nilsen          -  49u - 50u         9.63       9.21   
51        bjørnar     nilsen          -  36u - 33u         7.99       7.70   
134  kwemtochukwu  ogochukwu          -  34u - 37u         7.86       7.53   
264    chukwumoge  ogochukwu          -  41u - 39u         7.95       7.56   
194     eufrosina  schiavone          -  23u - 23u         9.00       8.56   
253        albino  schiavone  35u - 43u          -         7.56       7.15   
57         joseph     tucker  48u - 56u          -         7.67       7.30   
160        harley     tucker  47u - 57u          -         7.97       7.64   
22      beatrycze    woźniak          -  26u - 27u         7.54       7.17   
9            asia    woźniak  30u - 36u          -         7.76       7.37   

     hba1c_change  
19           0.34  
132           NaN  
49           0.38  
226          0.97  
199          0.36  
24           0.39  
219          0.35  
184          0.35  
13           0.34  
150          0.31  
10            NaN  
136           NaN  
39            NaN  
167          0.35  
238           NaN  
212          0.36  
66           0.95  
90           0.34  
183          0.36  
66           0.31  
236          0.37  
36            NaN  
3            0.92  
108          0.92  
213          0.92  
51            NaN  
134           NaN  
264          0.39  
194          0.94  
253           NaN  
57           0.37  
160          0.33  
22           0.37  
9             NaN

In [33]:
reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [34]:
reactions.isnull().sum()

given_name          0
surname             0
adverse_reaction    0
dtype: int64

In [35]:
reactions.duplicated().sum()

0

#### Assess:
   - `patients` table:
       - `patient_id` and `zip_code` should be str
       - `birthdate` is str
       - `weight` and `height` are in unfamiliar units
       - inaccurate value for `weight` for patient_id 211
       - inaccurate value for `height` for patient_id 5
       - `contact` columns has two data email and number
       - duplicated patient 'John Doe'
       - null values in columns `address`, `city`, `state`, `zip_code`, `country`, `contact`
       - duplication for patient_ids (25, 30), (98, 503), (132, 283)
       - in consistent names in `state` column
       - `given_name` and `surname` should be lower case
       
    
   - `treatments` table
       - column `hba1c_change` has missing values
       - columns `auralin` and `novodra` have two values start and end dose
       - patient 'joseph day' is duplicated
       - 
   - `reactions` table

## Clean

#### Define

#### Code

#### Test

## Visualization